daemon/graphdriver/copy/copy.go

```diff
@@ -15,26 +15,89 @@ import (
 	"golang.org/x/sys/unix"
 )
 
-type copyFlags int
+var copyFileRangeEnabled bool
+
+func init() {
+	copyFileRangeEnabled = isCopyFileRangeSyscallAvailable()
+}
+
+func isCopyFileRangeSyscallAvailable() bool {
+	if _, err := unix.CopyFileRange(-1, nil, -1, nil, 0, 0); err == unix.ENOSYS {
+		return false
+	}
+	return true
+}
+
+// The maximum amount of bytes Linux will splice
+const maxRwCount = 0x7ffff000
```

Comment:
Where is this documented? Copy file range takes in `size_t` which can be 64-bit http://man7.org/linux/man-pages/man2/copy_file_range.2.html.

Reply:
See: 
http://elixir.free-electrons.com/linux/v4.12.10/source/fs/read_write.c#L1621
http://elixir.free-electrons.com/linux/v4.12.10/source/include/linux/fs.h#L2138
`INT_MAX` is pretty much always based on 4-byte ints on Linux, on our currently supported platforms, and have 4096 byte pages.
 
For the majority of filesystems underneath us, we're going to be falling back to splice.

Reply:
In that case, we still don't need to define this variable here. We should just detect when less than the whole file was written and attempt the copy multiple times. However we can still pass in the full length and when clone is done there is no need to make multiple calls here.

Is this missing setting `ret = len;` here
http://elixir.free-electrons.com/linux/v4.12.10/source/fs/read_write.c#L1617